### ISSUE TITLE
Range.foreach is not specialised

### ISSUE THREAD

**Issue by limansky** [scala/scala3#19759]
## Compiler version

I've compared a simple peace of code performance running with Scala 2 and Scala 3 and faced with performance regression. In this case Scala 2 generated program is more than 7 times faster on my computer.
The code is just calculate sum of big Array[Long] several times.

## Minimized code

```Scala
    (0 until M).foreach { _ =>
      var i = 0
      while (i < N) {
        sum += arr(i)
        i += 1
      }
    }
```

The whole project code is available here:  https://github.com/limansky/perf-issue

## Output

```scala
sbt:perf> +run
[info] Setting Scala version to 2.13.12 on 1 projects.
[info] Reapplying settings...
[info] set current project to perf (in build file:/home/limansky/projects/langperf/scala/)
[info] running Main 
time 4268 ms
10000000000
[success] Total time: 5 s, completed Feb 22, 2024, 3:10:32 PM
[info] Setting Scala version to 3.3.1 on 1 projects.
[info] Reapplying settings...
[info] set current project to perf (in build file:/home/limansky/projects/langperf/scala/)
[info] running Main 
time 31699 ms
10000000000
```

## Expectation

I suppose Scala 3 should be at least as fast as Scala 2.

**Comment by sjrd** [comment]
Simple programs like that are not remotely good enough to reliably test performance on the JVM. Please use [JMH](https://github.com/openjdk/jmh) when attempting to compare run-time performance on the JVM.

**Comment by bishabosha** [comment]
actually the answer is very clear, Scala 2.13 specialised `foreach` to `foreach$mVc$sp`, but Scala 3 does not

**Comment by lrytz** [comment]
looking at cfr-decompiler output difference

```diff
         long s2 = System.nanoTime();
-        RichInt$.MODULE$.until$extension(Predef$.MODULE$.intWrapper(0), M).foreach$mVc$sp((Function1)(JFunction1.mcVI.sp & Serializable)x$1 -> {
+        RichInt$.MODULE$.until$extension(Predef$.MODULE$.intWrapper(0), M).foreach((Function1)(JFunction1.mcVI.sp & Serializable)_$1 -> {
             for (int i = 0; i < N; ++i) {
                 sum$1.elem += arr[i];
             }
```

[`Range.foreach` is specialized for `Int => Unit`](https://github.com/scala/scala/blob/v2.13.13/src/library/scala/collection/immutable/Range.scala#L184), but it seems Scala 3 doesn't invoke `foreach$mVc$sp`, so every int is boxed when calling the argument function.

**Comment by lrytz** [comment]
(jinx..)

**Comment by limansky** [comment]
Thanks, @bishabosha 
Replacing `foreach` with `while` makes performance the same. This is kind of counter-intuitive for me, because foreach is used for the outer loop, which has small number of iterations. Could you please explain why boxing of the outer loop variable has such significant performance impact in this case?

**Comment by som-snytt** [comment]
actually the clearer answer is that Scala 3 doesn't do `DelayedInit` anymore, so the example is compiled to an ordinary module which is not optimized i.e. jitted on the JVM because it is static initialization. (This is the old REPL bugaboo.)

Showing original in-module (the b version) vs factored out to instance method:
```
➜  snips scala i19759.scala
time 4148 ms
10000000000
➜  snips ~/projects/dotty/bin/scala i19759.scala
time 4118 ms
10000000000
➜  snips scala i19759b.scala
time 4184 ms
10000000000
➜  snips ~/projects/dotty/bin/scala i19759b.scala
time 70905 ms
10000000000
```
The original is the `b` version because when pasting I automatically moved it from `App` to regular class without thinking. That is what coding trauma does to the brain.

**Comment by bishabosha** [comment]
I removed the `regression` label because DelayedInit is explicitly a removed feature

**Comment by OndrejSpanel** [comment]
I have asked a  question in https://contributors.scala-lang.org/t/status-of-specialization-in-scala-3/4628 and my impression from the response was there is no support at all for specialization in Scala 3.

I know Scala 3 certainly cannot produce specialized code. Is it able to consume specialized functions created in Scala 2, or how can this issue be fixed?

**Comment by odersky** [comment]
@OndrejSpanel Yes, it should be able to consume specialized functions created in Scala 2. So it seems there's an issue with that specialization that needs fixing.

**Comment by bishabosha** [comment]
I did some experimentation on Saturday with a design where foreach is replaced with an `inline` method, and its not so simple:
1. you need to make the lambda parameter inline (for specialisation, and avoid closures), and then it is not allowed to override because of changing the parameter.
2. So then you can try and make an abstract inline foreach in the base Iterable trait, (because you cannot override a concrete inline) but that is suboptimal as then its illegal to call foreach on anything except concrete classes.

One example that can work is a single inline method in `Iterable` that can use inline match to specialise on the self type? (but it would mean renaming all existing `foreach` methods + overrides) (investigation in branch https://github.com/dotty-staging/dotty/tree/range-foreach-inline-investigation, and minimization in https://github.com/dotty-staging/dotty/commit/569c98b7f9698c2868fc59522b6b13f96a02e1b4)

second option - new language feature where an API designer can opt-into an inline extension method to be selected over a true method (if owner is companion object e.g.) - again this would only apply if static type is known to be Range - but if the method promises to have the same contract then why not (except that its a looser match because inline parameter would be allowed to substitute non-inline parameter)

third option: perhaps ability to introduce a marker to promise that the inline parameter will be evaluated exactly once, so can genuinely replace a non-inline parameter without refchecks complaining

backup option - just specialise the closures for unit and hope jit inlines the loop - shoudl check if there is a ceiling on this performance, hence why Spire introduced the `cfor` macro

**Comment by Syzygies** [comment]
I'm new to Scala, but I've used many languages including every functional language that Scala gets compared with. I have a language comparison project where I study the code and the performance of a dozen or so languages I might adopt for my ongoing math research. It's a toy problem tallying the cycle decompositions of 3,715,891,200 signed permutations on 10 letters in parallel, using Tarjan union-find and a faster loops algorithm.

I'm stunned that both the JVM and Native Scala code trounces OCaml and Haskell, up there near Rust with F#. I'm equally stunned that I find the actual Scala code more beautiful than Haskell or OCaml or Ruby or ... I'm a convert.

The one whopper I simply couldn't believe was this issue. In my case it isn't aesthetic or performance alone; my hot loops crashed the native allocator.

I had to write out tedious while loops, like I was back in my high school library using a paperclip to enter my first BASIC program into punched cards. This was extreme culture shock coming from Haskell monads. I went through the usual stages of grief, and now after benchmarking many alternatives, those same while loops beautiful to me.

I am head over heels in love with Scala, and I'm happy to crawl back under a rock and not care at all what you decide here. I just thought that another well-intended data point could help. I've met people who have visited 100 countries. I believe that I've tried as many languages. After adding Scala to a complex project, this issue was the single glaring wart in the entire experience that was otherwise overwhelmingly positive enough to turn me into a Scala fanatic.

You don't need test code to see that this should be fixed.
